rust/src/dns/dns.rs

```diff
@@ -115,6 +115,36 @@ pub const DNS_RCODE_BADTRUNC: u16 = 22;
 
 static mut ALPROTO_DNS: AppProto = ALPROTO_UNKNOWN;
 
+/// Trait for objects to implement methods that calculate the size of that object.
+pub trait SizeOf {
+    /// Returns the sizeof the object itself, plus the inner data.
+    fn sizeof(&self) -> usize {
+        std::mem::size_of_val(self) + self.sizeof_inner()
```

Comment:
This is awesome. :)
I have a question. How do we calculate the size of opaque type items with no explicit type from C that is used in tx? e.g. `DetectEngineState`

Reply:
You don't, at least not simply and accurately.  I see 2 options:
- Hardcode a size of known datatypes from C, these would have to be reviewed.
- Or define similar functions on the C side and call into them.
Or you just wrap it up into the extra overhead over and above the memcap that is likely to exist for any parser, and then memcap is close, but maybe 2-3% wrong.  At least within a specific Suricata version the margin of error will be constant (I don't think any of these C structs grow in size in response to specific network traffic).